File: app/src/main/java/com/kickstarter/libs/utils/ApplicationLifecycleUtil.java
package com.kickstarter.libs.utils;

import android.app.Activity;
import android.app.Application;
import android.content.ComponentCallbacks2;
import android.content.res.Configuration;
import android.os.Bundle;

import com.facebook.appevents.AppEventsLogger;
import com.kickstarter.KSApplication;
import com.kickstarter.libs.CurrentConfigType;
import com.kickstarter.libs.CurrentUserType;
import com.kickstarter.libs.Koala;
import com.kickstarter.libs.Logout;
import com.kickstarter.libs.qualifiers.KoalaTracker;
import com.kickstarter.libs.qualifiers.LakeTracker;
import com.kickstarter.libs.rx.transformers.Transformers;
import com.kickstarter.services.ApiClientType;
import com.kickstarter.services.apiresponses.ErrorEnvelope;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public final class ApplicationLifecycleUtil implements Application.ActivityLifecycleCallbacks, ComponentCallbacks2 {
  protected @Inject ApiClientType client;
  protected @Inject CurrentConfigType config;
  protected @Inject CurrentUserType currentUser;
  protected @Inject Logout logout;

  private final KSApplication application;
  private boolean isInBackground = true;

  public ApplicationLifecycleUtil(final @NonNull KSApplication application) {
    this.application = application;
    application.component().inject(this);
  }

  @Override
  public void onActivityCreated(final @NonNull Activity activity, final @Nullable Bundle bundle) {
  }

  @Override
  public void onActivityStarted(final @NonNull Activity activity) {
  }

  @Override
  public void onActivityResumed(final @NonNull Activity activity) {
    if(this.isInBackground){

      // Facebook: logs 'install' and 'app activate' App Events.
      AppEventsLogger.activateApp(activity.getApplication());

      // Refresh the config file
      this.client.config()
        .compose(Transformers.pipeApiErrorsTo(this::handleConfigApiError))
        .compose(Transformers.neverError())
        .subscribe(c -> this.config.config(c));

      // Refresh the user
      final String accessToken = this.currentUser.getAccessToken();

      if (ObjectUtils.isNotNull(accessToken)) {
        this.client.fetchCurrentUser()
          .compose(Transformers.neverError())
          .subscribe(u -> this.currentUser.refresh(u));
      }

      this.isInBackground = false;
    }
  }

  /**
   * Handles a config API error by logging the user out in the case of a 401. We will interpret
   * 401's on the config request as meaning the user's current access token is no longer valid,
   * as that endpoint should never 401 othewise.
   */
  private void handleConfigApiError(final @NonNull ErrorEnvelope error) {
    if (error.httpCode() == 401) {
      this.logout.execute();
      ApplicationUtils.startNewDiscoveryActivity(this.application);
    }
  }

  @Override
  public void onActivityPaused(final @NonNull Activity activity) {
    // Facebook: logs 'app deactivate' App Event.
    AppEventsLogger.deactivateApp(activity);
  }

  @Override
  public void onActivityStopped(final @NonNull Activity activity) {
  }

  @Override
  public void onActivitySaveInstanceState(final @NonNull Activity activity, final @Nullable Bundle bundle) {
  }

  @Override
  public void onActivityDestroyed(final @NonNull Activity activity) {
  }

  @Override
  public void onConfigurationChanged(final @NonNull Configuration configuration) {
  }

  @Override
  public void onLowMemory() {
  }

  /**
   * Memory availability callback. TRIM_MEMORY_UI_HIDDEN means the app's UI is no longer visible.
   * This is triggered when the user navigates out of the app and primarily used to free resources used by the UI.
   * http://developer.android.com/training/articles/memory.html
   */
  @Override
  public void onTrimMemory(final int i) {
    if(i == ComponentCallbacks2.TRIM_MEMORY_UI_HIDDEN) {
      this.isInBackground = true;
    }
  }
}

File: app/src/main/java/com/kickstarter/libs/utils/ApplicationLifecycleUtil.kt
package com.kickstarter.libs.utils

import android.app.Activity
import android.app.Application.ActivityLifecycleCallbacks
import android.content.ComponentCallbacks2
import android.content.res.Configuration
import android.os.Bundle
import com.facebook.appevents.AppEventsLogger
import com.google.firebase.analytics.FirebaseAnalytics
import com.kickstarter.KSApplication
import com.kickstarter.libs.Build
import com.kickstarter.libs.CurrentConfigTypeV2
import com.kickstarter.libs.CurrentUserTypeV2
import com.kickstarter.libs.Logout
import com.kickstarter.libs.preferences.StringPreferenceType
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.isNotNull
import com.kickstarter.libs.utils.extensions.syncUserFeatureFlagsFromPref
import com.kickstarter.services.ApiClientTypeV2
import com.kickstarter.services.apiresponses.ErrorEnvelope
import com.kickstarter.services.apiresponses.ErrorEnvelope.Companion.fromThrowable
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Inject

class ApplicationLifecycleUtil(private val application: KSApplication) :
    ActivityLifecycleCallbacks,
    ComponentCallbacks2 {

    @Inject
    lateinit var client: ApiClientTypeV2

    @Inject
    lateinit var config: CurrentConfigTypeV2

    @Inject
    lateinit var currentUser: CurrentUserTypeV2

    @Inject
    lateinit var logout: Logout

    @Inject
    lateinit var build: Build

    @JvmField
    @Inject
    var featuresFlagPreference: StringPreferenceType? = null
    private var isInBackground = true
    private val disposables = CompositeDisposable()

    init {
        application.component().inject(this)
    }

    override fun onActivityCreated(activity: Activity, bundle: Bundle?) {}
    override fun onActivityStarted(activity: Activity) {}
    override fun onActivityResumed(activity: Activity) {
        if (isInBackground) {
            // Facebook: logs 'install' and 'app activate' App Events.
            AppEventsLogger.activateApp(activity.application)
            refreshConfigFile()
            refreshUser()
            isInBackground = false
        }
    }

    /**
     * Refresh the config file.
     */
    private fun refreshConfigFile() {
        client.config()
            .materialize()
            .share()
            .subscribe { notification ->
                notification.value?.let {
                    // sync save features flags in the config object
                    if (build.isDebug || Build.isInternal()) {
                        it.syncUserFeatureFlagsFromPref(featuresFlagPreference!!)
                    }
                    config.config(it)
                }

                notification.error?.let {
                    handleConfigApiError(fromThrowable(it))
                }
            }.addToDisposable(disposables)
    }

    /**
     * Handles a config API error by logging the user out in the case of a 401. We will interpret
     * 401's on the config request as meaning the user's current access token is no longer valid,
     * as that endpoint should never 401 othewise.
     */
    private fun handleConfigApiError(error: ErrorEnvelope) {
        if (error.httpCode() == 401) {
            forceLogout("config_api_error")
        }
    }

    /**
     * Forces the current user session to be logged out.
     */
    private fun forceLogout(context: String) {
        logout.execute()
        ApplicationUtils.startNewDiscoveryActivity(application)
        val params = Bundle()
        params.putString("location", context)
        FirebaseAnalytics.getInstance(application).logEvent("force_logout", params)
    }

    /**
     * Refreshes the user object if there is not a user logged in with a non-null access token.
     */
    private fun refreshUser() {
        val accessToken = currentUser.accessToken ?: ""

        if (accessToken.isNotNull() && accessToken.isNotEmpty()) {
            client.fetchCurrentUser()
                .doOnError {
                    forceLogout(it.message ?: "")
                }
                .subscribe { user ->
                    currentUser.refresh(user)
                }.addToDisposable(disposables)
        }
    }

    override fun onActivityPaused(activity: Activity) { disposables.clear() }
    override fun onActivityStopped(activity: Activity) {}
    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
    override fun onActivityDestroyed(activity: Activity) {}
    override fun onConfigurationChanged(configuration: Configuration) {}
    override fun onLowMemory() {}

    /**
     * Memory availability callback. TRIM_MEMORY_UI_HIDDEN means the app's UI is no longer visible.
     * This is triggered when the user navigates out of the app and primarily used to free resources used by the UI.
     * http://developer.android.com/training/articles/memory.html
     */
    override fun onTrimMemory(i: Int) {
        if (i == ComponentCallbacks2.TRIM_MEMORY_UI_HIDDEN) {
            isInBackground = true
        }
    }
}
